Index: atk/Makefile.am
===================================================================
--- atk/Makefile.am	(revision 122401)
+++ atk/Makefile.am	(working copy)
@@ -22,4 +22,17 @@
 
 add_dist =
 
+	$(CSC) $(build_references) $(BUILD_PARAMS)
+
+MOON_BIN_PATH=../moonbin/
+
+references_moon = $(MOON_BIN_PATH)glib-sharp.dll
+build_references_moon = $(addprefix /r:, $(references_moon))
+
+moonlight: $(ASSEMBLY)
+	smcs -define:NET_2_1 $(build_references_moon) $(BUILD_PARAMS) -out:$(MOON_BIN_PATH)$(ASSEMBLY)
+	$(GAPI_CDECL_INSERT)
+
 include ../Makefile.include
Index: glib/Makefile.am
===================================================================
--- glib/Makefile.am	(revision 122401)
+++ glib/Makefile.am	(working copy)
@@ -97,11 +97,21 @@
 GAPI_CDECL_INSERT=
 endif
 
+BUILD_PARAMS=$(CSFLAGS) -unsafe -nowarn:0169,0612,0618 -target:library $(references) $(build_sources)
+
 $(ASSEMBLY): $(build_sources) gtk-sharp.snk AssemblyInfo.cs
 	@rm -f $(ASSEMBLY).mdb
-	$(CSC) $(CSFLAGS) -unsafe -nowarn:0169,0612,0618 -out:$(ASSEMBLY) -target:library $(references) $(build_sources)
+	$(CSC) $(BUILD_PARAMS) -out:$(ASSEMBLY)
 	$(GAPI_CDECL_INSERT)
 
+MOON_BIN_PATH=../moonbin/
+
+moonlight: $(ASSEMBLY)
+	rm -rf $(MOON_BIN_PATH)
+	mkdir $(MOON_BIN_PATH)
+	smcs -define:NET_2_1 $(BUILD_PARAMS) -out:$(MOON_BIN_PATH)$(ASSEMBLY)
+	$(GAPI_CDECL_INSERT)
+
 $(POLICY_ASSEMBLIES): $(top_builddir)/policy.config gtk-sharp.snk
 	@for i in $(POLICY_VERSIONS); do	\
 	  echo "Creating policy.$$i.$(ASSEMBLY)";	\
Index: Makefile.include
===================================================================
--- Makefile.include	(revision 122401)
+++ Makefile.include	(working copy)
@@ -62,9 +62,11 @@
 GAPI_CDECL_INSERT=
 endif
 
+BUILD_PARAMS=$(CSFLAGS) -nowarn:0169,0612,0618 -unsafe -target:library $(GENERATED_SOURCES) $(build_sources)
+
 $(ASSEMBLY): generated-stamp $(SNK) $(build_sources) $(references)
 	@rm -f $(ASSEMBLY).mdb
-	$(CSC) $(CSFLAGS) -nowarn:0169,0612,0618 -unsafe -out:$(ASSEMBLY) -target:library $(build_references) $(GENERATED_SOURCES) $(build_sources)
+	$(CSC) $(build_references) $(BUILD_PARAMS) -out:$(ASSEMBLY)
 	$(GAPI_CDECL_INSERT)
 
 install-data-local:
