Index: hello.c
===================================================================
--- hello.c	(revision 98482)
+++ hello.c	(working copy)
@@ -11,7 +11,7 @@
 gboolean quit(GIOChannel *source, GIOCondition condition, gpointer data)
 {
   g_signal_emit_by_name(root, "children-changed::remove", child);
-  g_object_unref(child);
+  if (child) g_object_unref(child);
   child = NULL;
   g_object_unref(root);
   tcsetattr(0, TCSAFLUSH, data);
@@ -59,7 +59,7 @@
   printf("ref child: %p, index %d\n", obj, i);
   if (obj == root && i == 0)
   {
-    g_object_ref(child);
+    if (child) g_object_ref(child);
     return child;
   }
   else return NULL;
@@ -136,21 +136,51 @@
   return root;
 }
 
+static G_CONST_RETURN gchar *get_toolkit_name()
+{
+  return "hello";
+}
+
+static G_CONST_RETURN gchar *get_toolkit_version()
+{
+  return "1.0";
+}
+
 main(int argc, char *argv[])
 {
   AtkUtilClass *klass;
   GMainLoop *mainloop;
   GIOChannel *stdin_channel;
   struct termios termios, rt;
+  GModule *bridge;
+  void (*gnome_accessibility_module_init)();
 
-  /* The following is normally done by Gnome if accessibility is enabled */
-  putenv("GTK_MODULES=gail:atk-bridge");
-  
-  gtk_init(&argc, &argv);
+  g_type_init();
+
   klass = g_type_class_ref(ATK_TYPE_UTIL);
   klass->get_root = get_root;
+  klass->get_toolkit_name = get_toolkit_name;
+  klass->get_toolkit_version = get_toolkit_version;
   g_type_class_unref(klass);
+
   root = g_object_new(TEST_TYPE_HELLO, NULL);
+
+  /* The below line isn't really right -- normally gtk will build the path */
+  bridge = g_module_open("/usr/lib/gtk-2.0/modules/libatk-bridge.so", G_MODULE_BIND_LOCAL|G_MODULE_BIND_LAZY);
+
+  if (!bridge)
+  {
+    fprintf(stderr, "Couldn't load atk-bridge.so\n");
+    exit(1);
+  }
+
+  if (!g_module_symbol(bridge, "gnome_accessibility_module_init", (gpointer *)&gnome_accessibility_module_init))
+  {
+    fprintf(stderr, "Couldn't find gnome_accessibility_module_init\n");    
+    exit(1);
+  }
+
+  (*gnome_accessibility_module_init)();
   atk_object_set_name(root, "root object");
 
   //NOTE: I comment this in order to have the smallest testcase that works
Index: hello.h
===================================================================
--- hello.h	(revision 98177)
+++ hello.h	(working copy)
@@ -1,5 +1,5 @@
 #ifndef TEST_HELLO_H
-#include "gtk/gtk.h"
+#include "gmodule.h"
 #include <atk/atk.h>
 
 G_BEGIN_DECLS
Index: atkHelloWorld.mdp
===================================================================
--- atkHelloWorld.mdp	(revision 98257)
+++ atkHelloWorld.mdp	(working copy)
@@ -21,6 +21,6 @@
   <compiler ctype="GccCompiler" />
   <Packages>
     <ProjectPackage file="atk" name="atk" IsProject="False" />
-    <ProjectPackage file="gtk+-2.0" name="gtk+-2.0" IsProject="False" />
+    <ProjectPackage file="gmodule" name="gmodule" IsProject="False" />
   </Packages>
 </Project>
\ No newline at end of file
