Index: gnome-sharp/gconf/tools/gconfsharp2-schemagen.in
===================================================================
--- gnome-sharp/gconf/tools/gconfsharp2-schemagen.in	(revision 76653)
+++ gnome-sharp/gconf/tools/gconfsharp2-schemagen.in	(working copy)
@@ -1,6 +1,2 @@
 #!/bin/sh
-
-a=`which "$0"`
-d=`dirname "$a"`
-
-@RUNTIME@ $d/../lib/gtk-sharp-2.0/gconfsharp-schemagen.exe "$@"
+@RUNTIME@ @prefix@/lib/gtk-sharp-2.0/gconfsharp-schemagen.exe "$@"
