Index: gtk-sharp/generator/gapi2-codegen.in
===================================================================
--- gtk-sharp/generator/gapi2-codegen.in	(revision 76652)
+++ gtk-sharp/generator/gapi2-codegen.in	(working copy)
@@ -1,6 +1,2 @@
 #!/bin/sh
-
-a=`which "$0"`
-d=`dirname "$a"`
-
-@RUNTIME@ $d/../lib/gtk-sharp-2.0/gapi_codegen.exe "$@"
+@RUNTIME@ @prefix@/lib/gtk-sharp-2.0/gapi_codegen.exe "$@"
Index: gtk-sharp/parser/gapi2-fixup.in
===================================================================
--- gtk-sharp/parser/gapi2-fixup.in	(revision 76652)
+++ gtk-sharp/parser/gapi2-fixup.in	(working copy)
@@ -1,6 +1,2 @@
 #!/bin/sh
-
-a=`which "$0"`
-d=`dirname "$a"`
-
-@RUNTIME@ $d/../lib/gtk-sharp-2.0/gapi-fixup.exe "$@"
+@RUNTIME@ @prefix@/lib/gtk-sharp-2.0/gapi-fixup.exe "$@"
Index: gtk-sharp/parser/gapi2-parser.in
===================================================================
--- gtk-sharp/parser/gapi2-parser.in	(revision 76652)
+++ gtk-sharp/parser/gapi2-parser.in	(working copy)
@@ -1,7 +1,3 @@
 #!/bin/sh
-
-a=`which "$0"`
-d=`dirname "$a"`
-
-export PATH=$d/../lib/gtk-sharp-2.0:$PATH
-@RUNTIME@ $d/../lib/gtk-sharp-2.0/gapi-parser.exe "$@"
+export PATH=@prefix@/lib/gtk-sharp-2.0:$PATH
+@RUNTIME@ @prefix@/lib/gtk-sharp-2.0/gapi-parser.exe "$@"
