diff -uraN mono-uia-1.0.orig/data/Makefile.am mono-uia-1.0/data/Makefile.am
--- mono-uia-1.0.orig/data/Makefile.am	2009-05-21 01:10:48.518263417 +0200
+++ mono-uia-1.0/data/Makefile.am	2009-05-21 01:13:36.248263285 +0200
@@ -1,8 +1,8 @@
 
 EXTRA_DIST =  
 
-net20dir   = $(prefix)/lib/mono/2.0
-a11ydir    = $(prefix)/lib/mono/accessibility
+net20dir   = $(libdir)/mono/2.0
+a11ydir    = $(libdir)/mono/accessibility
 
 SHELL = /bin/sh
 MKINSTALLDIRS = $(SHELL) $(top_srcdir)/mkinstalldirs
diff -uraN mono-uia-1.0.orig/data/Makefile.in mono-uia-1.0/data/Makefile.in
--- mono-uia-1.0.orig/data/Makefile.in	2009-05-21 01:10:48.518263417 +0200
+++ mono-uia-1.0/data/Makefile.in	2009-05-21 01:14:07.038263566 +0200
@@ -127,8 +127,8 @@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
 EXTRA_DIST = 
-net20dir = $(prefix)/lib/mono/2.0
-a11ydir = $(prefix)/lib/mono/accessibility
+net20dir = $(libdir)/mono/2.0
+a11ydir = $(libdir)/mono/accessibility
 MKINSTALLDIRS = $(SHELL) $(top_srcdir)/mkinstalldirs
 INSTALL_D = $(INSTALL) -c -m 644
 REWRITE = 
diff -uraN mono-uia-1.0.orig/Makefile.include mono-uia-1.0/Makefile.include
--- mono-uia-1.0.orig/Makefile.include	2009-05-21 01:10:48.513263283 +0200
+++ mono-uia-1.0/Makefile.include	2009-05-21 01:13:07.307263280 +0200
@@ -107,7 +107,7 @@
 	mkdir -p '$(@D)'
 	$(AL) -out:'$@' -culture:$* -t:lib $(cmd_line_satellite_$*)
 
-GACROOT = $(DESTDIR)$(prefix)/lib
+GACROOT = $(DESTDIR)$(libdir)
 
 gac-install: $(POLICY_ASSEMBLIES)
 	$(GACUTIL) /i $(ASSEMBLY) /f $(GACUTIL_FLAGS) /root $(GACROOT) $(PACKAGE_FLAG) || exit 1;
diff -uraN mono-uia-1.0.orig/UIAutomationBridge/Makefile.in mono-uia-1.0/UIAutomationBridge/Makefile.in
--- mono-uia-1.0.orig/UIAutomationBridge/Makefile.in	2009-05-21 01:10:48.555261784 +0200
+++ mono-uia-1.0/UIAutomationBridge/Makefile.in	2009-05-21 01:15:22.878016069 +0200
@@ -245,7 +245,7 @@
 culture_resource_commandlines = $(call unesc2,cmd_line_satellite_$1 += '/embed:$(subst .resx,.resources,$2)')
 build_satellite_assembly_list = $(call q2s,$(cultures:%=$(BUILD_DIR)/%/$(SATELLITE_ASSEMBLY_NAME)))
 build_culture_res_files = $(call q2s,$(foreach res, $(culture_resources),$(call get_resource_name,$(res))))
-GACROOT = $(DESTDIR)$(prefix)/lib
+GACROOT = $(DESTDIR)$(libdir)
 all: all-am
 
 .SUFFIXES:
diff -uraN mono-uia-1.0.orig/UIAutomationClient/Makefile.in mono-uia-1.0/UIAutomationClient/Makefile.in
--- mono-uia-1.0.orig/UIAutomationClient/Makefile.in	2009-05-21 01:10:48.544262732 +0200
+++ mono-uia-1.0/UIAutomationClient/Makefile.in	2009-05-21 01:14:50.869264343 +0200
@@ -251,7 +251,7 @@
 culture_resource_commandlines = $(call unesc2,cmd_line_satellite_$1 += '/embed:$(subst .resx,.resources,$2)')
 build_satellite_assembly_list = $(call q2s,$(cultures:%=$(BUILD_DIR)/%/$(SATELLITE_ASSEMBLY_NAME)))
 build_culture_res_files = $(call q2s,$(foreach res, $(culture_resources),$(call get_resource_name,$(res))))
-GACROOT = $(DESTDIR)$(prefix)/lib
+GACROOT = $(DESTDIR)$(libdir)
 all: all-am
 
 .SUFFIXES:
diff -uraN mono-uia-1.0.orig/UIAutomationProvider/Makefile.in mono-uia-1.0/UIAutomationProvider/Makefile.in
--- mono-uia-1.0.orig/UIAutomationProvider/Makefile.in	2009-05-21 01:10:48.540262359 +0200
+++ mono-uia-1.0/UIAutomationProvider/Makefile.in	2009-05-21 01:14:33.376263360 +0200
@@ -275,7 +275,7 @@
 culture_resource_commandlines = $(call unesc2,cmd_line_satellite_$1 += '/embed:$(subst .resx,.resources,$2)')
 build_satellite_assembly_list = $(call q2s,$(cultures:%=$(BUILD_DIR)/%/$(SATELLITE_ASSEMBLY_NAME)))
 build_culture_res_files = $(call q2s,$(foreach res, $(culture_resources),$(call get_resource_name,$(res))))
-GACROOT = $(DESTDIR)$(prefix)/lib
+GACROOT = $(DESTDIR)$(libdir)
 all: all-am

 .SUFFIXES:
diff -uraN mono-uia-1.0.orig/UIAutomationHelpers/Makefile.in mono-uia-1.0/UIAutomationHelpers/Makefile.in
--- mono-uia-1.0.orig/UIAutomationHelpers/Makefile.in	2009-05-21 01:10:48.553261633 +0200
+++ mono-uia-1.0/UIAutomationHelpers/Makefile.in	2009-05-21 01:15:06.158289227 +0200
@@ -351,7 +351,7 @@
 culture_resource_commandlines = $(call unesc2,cmd_line_satellite_$1 += '/embed:$(subst .resx,.resources,$2)')
 build_satellite_assembly_list = $(call q2s,$(cultures:%=$(BUILD_DIR)/%/$(SATELLITE_ASSEMBLY_NAME)))
 build_culture_res_files = $(call q2s,$(foreach res, $(culture_resources),$(call get_resource_name,$(res))))
-GACROOT = $(DESTDIR)$(prefix)/lib
+GACROOT = $(DESTDIR)$(libdir)
 all: all-am

 .SUFFIXES:
diff -uraN mono-uia-1.0.orig/UIAutomationTypes/Makefile.in mono-uia-1.0/UIAutomationTypes/Makefile.in
--- mono-uia-1.0.orig/UIAutomationTypes/Makefile.in	2009-05-21 01:10:48.553261633 +0200
+++ mono-uia-1.0/UIAutomationTypes/Makefile.in	2009-05-21 01:15:06.158289227 +0200
@@ -301,7 +301,7 @@
 culture_resource_commandlines = $(call unesc2,cmd_line_satellite_$1 += '/embed:$(subst .resx,.resources,$2)')
 build_satellite_assembly_list = $(call q2s,$(cultures:%=$(BUILD_DIR)/%/$(SATELLITE_ASSEMBLY_NAME)))
 build_culture_res_files = $(call q2s,$(foreach res, $(culture_resources),$(call get_resource_name,$(res))))
-GACROOT = $(DESTDIR)$(prefix)/lib
+GACROOT = $(DESTDIR)$(libdir)
 all: all-am
 
 .SUFFIXES:
diff -uraN mono-uia-1.0.orig/WindowsBase/Makefile.in mono-uia-1.0/WindowsBase/Makefile.in
--- mono-uia-1.0.orig/WindowsBase/Makefile.in	2009-05-21 01:10:48.511034894 +0200
+++ mono-uia-1.0/WindowsBase/Makefile.in	2009-05-21 01:12:46.888263605 +0200
@@ -245,7 +245,7 @@
 culture_resource_commandlines = $(call unesc2,cmd_line_satellite_$1 += '/embed:$(subst .resx,.resources,$2)')
 build_satellite_assembly_list = $(call q2s,$(cultures:%=$(BUILD_DIR)/%/$(SATELLITE_ASSEMBLY_NAME)))
 build_culture_res_files = $(call q2s,$(foreach res, $(culture_resources),$(call get_resource_name,$(res))))
-GACROOT = $(DESTDIR)$(prefix)/lib
+GACROOT = $(DESTDIR)$(libdir)
 all: all-am
 
 .SUFFIXES:
diff -uraN UIAutomation/UIAutomationSource/Makefile.in.old UIAutomation/UIAutomationSource/Makefile.in
--- UIAutomation/UIAutomationSource/Makefile.in.old	2010-01-11 15:09:01.000000000 -0700
+++ UIAutomation/UIAutomationSource/Makefile.in	2010-01-11 15:07:17.000000000 -0700
@@ -281,7 +281,7 @@
 culture_resource_commandlines = $(call unesc2,cmd_line_satellite_$1 += '/embed:$(subst .resx,.resources,$2)')
 build_satellite_assembly_list = $(call q2s,$(cultures:%=$(BUILD_DIR)/%/$(SATELLITE_ASSEMBLY_NAME)))
 build_culture_res_files = $(call q2s,$(foreach res, $(culture_resources),$(call get_resource_name,$(res))))
-GACROOT = $(DESTDIR)$(prefix)/lib
+GACROOT = $(DESTDIR)$(libdir)
 all: all-am
 
 .SUFFIXES:
