Index: glib/Signal.cs
===================================================================
--- glib/Signal.cs	(revision 102534)
+++ glib/Signal.cs	(working copy)
@@ -295,11 +295,23 @@
 			GLib.Value ret = GLib.Value.Empty;
 
 			g_signal_emitv (inst_and_params.ArrayPtr, signal_id, gquark, ref ret);
-			object ret_obj = ret.Val;
+			object ret_obj = null;
+			try {
+				// may throw:
+				// Unhandled Exception: System.Exception: Unknown type
+				//  at GLib.Value.get_Val () [0x00000]
+				ret_obj = ret.Val;
+				
+				// if we dispose being an invalid value:
+				// GLib-GObject-CRITICAL **: g_value_unset: assertion `G_IS_VALUE (value)' failed
+				ret.Dispose ();
+			}
+			catch
+			{
+			}
 			
 			foreach (GLib.Value val in vals)
 				val.Dispose ();
-			ret.Dispose ();
 			
 			return ret_obj;
 		}
