Index: class/Managed.Windows.Forms/System.Windows.Forms/ComboBox.cs
===================================================================
--- class/Managed.Windows.Forms/System.Windows.Forms/ComboBox.cs	(revision 108175)
+++ class/Managed.Windows.Forms/System.Windows.Forms/ComboBox.cs	(working copy)
@@ -1913,7 +1913,23 @@
 
 			private ComboBox owner;
 			internal ArrayList object_items = new ArrayList ();
+			
+			#region UIA Framework Events
 
+#if NET_2_0
+	    		//UIA Framework events for ComboBox Provider used the
+	    		//generated UIA StructureChanged event
+	    		//- ChildAdded. Generates StructureChanged.ChildAdded
+	    		//- ChildRemoved. Generates StructureChanged.ChildRemoved
+	    		//- ChildrenClear. Generates StructureChanged.ChildrenReordered	    		
+			internal event StructureChangeEventHandler ChildAdded;
+			internal event StructureChangeEventHandler ChildRemoved;
+			internal event EventHandler ChildrenClear;
+			internal delegate void StructureChangeEventHandler (object sender, int index);
+#endif
+
+			#endregion
+
 			public ObjectCollection (ComboBox owner)
 			{
 				this.owner = owner;
@@ -1943,7 +1959,22 @@
 					if (value == null)
 						throw new ArgumentNullException ("value");
 
+#if NET_2_0
+					//UIA Framework event used to indicate that Structure
+					//has Changed: child item removed
+					if (ChildRemoved != null)
+						ChildRemoved (owner, index);
+#endif
+
 					object_items[index] = value;
+					
+#if NET_2_0
+					//UIA Framework event used to indicate that Structure
+					//has Changed: child item added
+					if (ChildAdded != null)
+						ChildAdded (owner, index);
+#endif
+
 					if (owner.listbox_ctrl != null)
 						owner.listbox_ctrl.InvalidateItem (index);
 					if (index == owner.SelectedIndex) {
@@ -1996,6 +2027,13 @@
 				object_items.Clear ();
 				owner.UpdatedItems ();
 				owner.Refresh ();
+				
+#if NET_2_0
+				//UIA Framework event used to indicate that Structure
+				//has Changed: children clear
+				if (ChildrenClear != null)
+					ChildrenClear (owner, EventArgs.Empty);
+#endif
 			}
 			
 			public bool Contains (object value)
@@ -2057,8 +2095,15 @@
 				
 				if (owner.Sorted)
 					AddItem (item);
-				else
+				else {
 					object_items.Insert (index, item);
+#if NET_2_0
+					//UIA Framework event used to indicate that Structure
+					//has Changed: child item added
+					if (ChildAdded != null)
+						ChildAdded (owner, index);
+#endif					
+				}
 				
 				owner.EndUpdate ();	// Calls UpdatedItems
 			}
@@ -2084,6 +2129,13 @@
 
 				object_items.RemoveAt (index);
 				owner.UpdatedItems ();
+				
+#if NET_2_0
+				//UIA Framework event used to indicate that Structure
+				//has Changed: child item removed
+				if (ChildRemoved != null)
+					ChildRemoved (owner, index);
+#endif
 			}
 			#endregion Public Methods
 
@@ -2105,12 +2157,27 @@
 							if (index <= owner.selected_index && owner.IsHandleCreated)
 								owner.selected_index++;
 								
+#if NET_2_0
+							//UIA Framework event used to indicate that Structure
+							//has Changed: child item added
+							if (ChildAdded != null)
+								ChildAdded (owner, index);
+#endif
+
 							return index;
 						}
 						index++;
 					}
 				}
 				object_items.Add (item);
+				
+#if NET_2_0
+				//UIA Framework event used to indicate that Structure
+				//has Changed: child item added
+				if (ChildAdded != null)
+					ChildAdded (owner, object_items.Count - 1);
+#endif
+				
 				return object_items.Count - 1;
 			}
 			
