Index: class/Managed.Windows.Forms/System.Windows.Forms/ComboBox.cs
===================================================================
--- class/Managed.Windows.Forms/System.Windows.Forms/ComboBox.cs	(revision 111121)
+++ class/Managed.Windows.Forms/System.Windows.Forms/ComboBox.cs	(working copy)
@@ -1913,7 +1913,25 @@
 
 			private ComboBox owner;
 			internal ArrayList object_items = new ArrayList ();
+			
+			#region UIA Framework Events
 
+#if NET_2_0
+			//NOTE:
+			//	We are using Reflection to add/remove internal events.
+			//	Class ListProvider uses the events.
+			//
+			//Event used to generate UIA StructureChangedEvent
+			static object CollectionChangedEvent = new object ();
+
+			internal event CollectionChangeEventHandler CollectionChanged {
+				add { owner.Events.AddHandler (CollectionChangedEvent, value); }
+				remove { owner.Events.RemoveHandler (CollectionChangedEvent, value); }
+			}
+#endif
+
+			#endregion
+
 			public ObjectCollection (ComboBox owner)
 			{
 				this.owner = owner;
@@ -1943,7 +1961,26 @@
 					if (value == null)
 						throw new ArgumentNullException ("value");
 
+#if NET_2_0
+					//UIA Framework event: Item Removed
+					CollectionChangeEventHandler eh
+						= (CollectionChangeEventHandler) owner.Events [CollectionChangedEvent];
+					if (eh != null)
+						eh (owner,
+						    new CollectionChangeEventArgs (CollectionChangeAction.Remove,
+						                                   index));
+#endif
+
 					object_items[index] = value;
+					
+#if NET_2_0
+					//UIA Framework event: Item Added
+					if (eh != null)
+						eh (owner,
+						    new CollectionChangeEventArgs (CollectionChangeAction.Add,
+						                                   index));
+#endif
+
 					if (owner.listbox_ctrl != null)
 						owner.listbox_ctrl.InvalidateItem (index);
 					if (index == owner.SelectedIndex) {
@@ -1996,6 +2033,16 @@
 				object_items.Clear ();
 				owner.UpdatedItems ();
 				owner.Refresh ();
+				
+#if NET_2_0
+				//UIA Framework event: Items list cleared
+				CollectionChangeEventHandler eh
+					= (CollectionChangeEventHandler) owner.Events [CollectionChangedEvent];
+				if (eh != null)
+					eh (owner,
+					    new CollectionChangeEventArgs (CollectionChangeAction.Refresh,
+					                                   -1));
+#endif
 			}
 			
 			public bool Contains (object value)
@@ -2057,8 +2104,18 @@
 				
 				if (owner.Sorted)
 					AddItem (item);
-				else
+				else {
 					object_items.Insert (index, item);
+#if NET_2_0
+					//UIA Framework event: Item added
+					CollectionChangeEventHandler eh
+						= (CollectionChangeEventHandler) owner.Events [CollectionChangedEvent];
+					if (eh != null)
+						eh (owner,
+						    new CollectionChangeEventArgs (CollectionChangeAction.Add,
+						                                   index));
+#endif					
+				}
 				
 				owner.EndUpdate ();	// Calls UpdatedItems
 			}
@@ -2084,6 +2141,16 @@
 
 				object_items.RemoveAt (index);
 				owner.UpdatedItems ();
+				
+#if NET_2_0
+				//UIA Framework event: Item removed
+				CollectionChangeEventHandler eh
+					= (CollectionChangeEventHandler) owner.Events [CollectionChangedEvent];
+				if (eh != null)
+					eh (owner,
+					    new CollectionChangeEventArgs (CollectionChangeAction.Remove,
+					                                   index));
+#endif
 			}
 			#endregion Public Methods
 
@@ -2105,12 +2172,33 @@
 							if (index <= owner.selected_index && owner.IsHandleCreated)
 								owner.selected_index++;
 								
+#if NET_2_0
+							//UIA Framework event: Item added
+							CollectionChangeEventHandler eh
+								= (CollectionChangeEventHandler) owner.Events [CollectionChangedEvent];
+							if (eh != null)
+								eh (owner,
+								    new CollectionChangeEventArgs (CollectionChangeAction.Add,
+					        			                           index));
+#endif
+
 							return index;
 						}
 						index++;
 					}
 				}
 				object_items.Add (item);
+				
+#if NET_2_0
+				//UIA Framework event: Item added
+				CollectionChangeEventHandler eha
+					= (CollectionChangeEventHandler) owner.Events [CollectionChangedEvent];
+				if (eha != null)
+					eha (owner,
+					     new CollectionChangeEventArgs (CollectionChangeAction.Add,
+					    	                            object_items.Count - 1));
+#endif
+				
 				return object_items.Count - 1;
 			}
 			
