Index: class/Managed.Windows.Forms/System.Windows.Forms/ListBox.cs
===================================================================
--- class/Managed.Windows.Forms/System.Windows.Forms/ListBox.cs	(revision 110245)
+++ class/Managed.Windows.Forms/System.Windows.Forms/ListBox.cs	(working copy)
@@ -2383,6 +2383,25 @@
 
 			private ListBox owner;
 			internal ArrayList object_items = new ArrayList ();
+			
+			#region UIA Framework Events 
+#if NET_2_0
+			//NOTE:
+			//	We are using Reflection to add/remove internal events.
+			//	Class ListProvider uses the events.
+			//
+	    		//Events used to generate UIA StructureChangedEvent
+	    		// * ChildAdded: Item has been added.
+	    		// * ChildRemoved: Item has been removed.
+	    		// * ChildrenCleared: Items list has been cleared.
+	    		// * ChildrenSorted: Items list has been sorted.
+			internal event StructureChangeEventHandler ChildAdded;
+			internal event StructureChangeEventHandler ChildRemoved;
+			internal event EventHandler ChildrenCleared;
+			internal event EventHandler ChildrenSorted;
+			internal delegate void StructureChangeEventHandler (object sender, int index);
+#endif
+			#endregion
 
 			public ObjectCollection (ListBox owner)
 			{
@@ -2424,8 +2443,22 @@
 						throw new ArgumentOutOfRangeException ("Index of out range");
 					if (value == null)
 						throw new ArgumentNullException ("value");
+						
+#if NET_2_0
+					//UIA Framework event: Item Removed
+					if (ChildRemoved != null)
+						ChildRemoved (owner, index);
+#endif
 
 					object_items[index] = value;
+					
+#if NET_2_0
+					//UIA Framework event: Item Added
+					if (ChildAdded != null)
+						ChildAdded (owner, index);
+#endif					
+
+					
 					owner.CollectionChanged ();
 				}
 			}
@@ -2489,6 +2522,12 @@
 				owner.selected_indices.Clear ();
 				object_items.Clear ();
 				owner.CollectionChanged ();
+
+#if NET_2_0
+				//UIA Framework event: Items list cleared
+				if (ChildrenCleared != null)
+					ChildrenCleared (owner, EventArgs.Empty);
+#endif
 			}
 			public bool Contains (object value)
 			{
@@ -2549,6 +2588,12 @@
 				object_items.Insert (index, item);
 				owner.CollectionChanged ();
 				owner.EndUpdate ();
+				
+#if NET_2_0
+				//UIA Framework event: Item Added
+				if (ChildAdded != null)
+					ChildAdded (owner, index);
+#endif
 			}
 
 			public void Remove (object value)
@@ -2567,6 +2612,12 @@
 				owner.selected_indices.Remove (index);
 				object_items.RemoveAt (index);
 				owner.CollectionChanged ();
+				
+#if NET_2_0
+				//UIA Framework event: Item Removed
+				if (ChildRemoved != null)
+					ChildRemoved (owner, index);
+#endif
 			}
 			#endregion Public Methods
 
@@ -2578,12 +2629,26 @@
 
 				int cnt = object_items.Count;
 				object_items.Add (item);
+
+#if NET_2_0
+				//UIA Framework event: Item Added
+				if (ChildAdded != null)
+					ChildAdded (owner, cnt);
+#endif
+
 				return cnt;
 			}
 
 			internal void Sort ()
 			{
 				object_items.Sort (new ListObjectComparer ());
+
+#if NET_2_0
+				//UIA Framework event: Items list Sorted
+				if (ChildrenSorted != null)
+					ChildrenSorted (owner, EventArgs.Empty);
+#endif
+
 			}
 
 			#endregion Private Methods
