Index: class/Managed.Windows.Forms/System.Windows.Forms/ListView.cs
===================================================================
--- class/Managed.Windows.Forms/System.Windows.Forms/ListView.cs	(revision 116548)
+++ class/Managed.Windows.Forms/System.Windows.Forms/ListView.cs	(working copy)
@@ -762,7 +762,14 @@
 		[DefaultValue (true)]
 		public bool MultiSelect {
 			get { return multiselect; }
-			set { multiselect = value; }
+			set {
+				multiselect = value; 
+
+#if NET_2_0
+				// UIA Framework: Event used by Selection Pattern in ListView provider
+				OnUIAMultiSelectChanged ();
+#endif
+			}
 		}
 
 
@@ -1046,6 +1053,11 @@
 					h_scroll.Value = v_scroll.Value = 0;
 					view = value; 
 					Redraw (true);
+
+#if NET_2_0
+					// UIA Framework: Event used to update UIA Tree.
+					OnUIAViewChanged ();
+#endif
 				}
 			}
 		}
@@ -4909,6 +4921,31 @@
 			private ListViewGroup group;
 #endif
 
+			#region UIA Framework Events 
+#if NET_2_0
+			//NOTE:
+			//	We are using Reflection to add/remove internal events.
+			//	Class ListViewProvider uses the events.
+			//
+			//Event used to generate UIA StructureChangedEvent
+			static object UIACollectionChangedEvent = new object ();
+
+			internal event CollectionChangeEventHandler UIACollectionChanged {
+				add { owner.Events.AddHandler (UIACollectionChangedEvent, value); }
+				remove { owner.Events.RemoveHandler (UIACollectionChangedEvent, value); }
+			}
+
+			internal void OnUIACollectionChangedEvent (CollectionChangeEventArgs args)
+			{
+				CollectionChangeEventHandler eh
+					= (CollectionChangeEventHandler) owner.Events [UIACollectionChangedEvent];
+				if (eh != null)
+					eh (owner, args);
+			}
+
+#endif
+			#endregion UIA Framework Events 
+
 			// The collection can belong to a ListView (main) or to a ListViewGroup (sub-collection)
 			// In the later case ListViewItem.ListView never gets modified
 			private bool is_main_collection = true;
@@ -4996,7 +5033,18 @@
 					}
 #endif
 
+#if NET_2_0
+					//UIA Framework event: Item Removed
+					OnUIACollectionChangedEvent (new CollectionChangeEventArgs (CollectionChangeAction.Remove, list [index]));
+#endif
+
 					list [index] = value;
+
+#if NET_2_0
+					//UIA Framework event: Item Removed
+					OnUIACollectionChangedEvent (new CollectionChangeEventArgs (CollectionChangeAction.Add, value));
+#endif
+
 					CollectionChanged (true);
 				}
 			}
@@ -5028,10 +5076,20 @@
 			object IList.this [int index] {
 				get { return this [index]; }
 				set {
+#if NET_2_0
+					//UIA Framework event: Item Replaced
+					OnUIACollectionChangedEvent (new CollectionChangeEventArgs (CollectionChangeAction.Remove, this [index]));
+#endif
+
 					if (value is ListViewItem)
 						this [index] = (ListViewItem) value;
 					else
 						this [index] = new ListViewItem (value.ToString ());
+#if NET_2_0
+					//UIA Framework event: Item Replaced
+					OnUIACollectionChangedEvent (new CollectionChangeEventArgs (CollectionChangeAction.Add, value));
+#endif
+
 					OnChange ();
 				}
 			}
@@ -5149,6 +5207,12 @@
 
 				list.Clear ();
 				CollectionChanged (false);
+
+#if NET_2_0
+				//UIA Framework event: Items Removed
+				OnUIACollectionChangedEvent (new CollectionChangeEventArgs (CollectionChangeAction.Refresh, null));
+#endif
+
 			}
 
 			public bool Contains (ListViewItem item)
@@ -5228,6 +5292,11 @@
 				result = list.Add (li);
 				CollectionChanged (true);
 
+#if NET_2_0
+				//UIA Framework event: Item Added
+				OnUIACollectionChangedEvent (new CollectionChangeEventArgs (CollectionChangeAction.Add, li));
+#endif
+
 				return result;
 			}
 
@@ -5247,6 +5316,11 @@
 					this.Insert (index, (ListViewItem) item);
 				else
 					this.Insert (index, item.ToString ());
+
+#if NET_2_0
+				//UIA Framework event: Item Added
+				OnUIACollectionChangedEvent (new CollectionChangeEventArgs (CollectionChangeAction.Add, this [index]));
+#endif
 			}
 
 			void IList.Remove (object item)
@@ -5317,6 +5391,11 @@
 				if (is_main_collection || item.ListView != null)
 					CollectionChanged (true);
 
+#if NET_2_0
+				//UIA Framework event: Item Added
+				OnUIACollectionChangedEvent (new CollectionChangeEventArgs (CollectionChangeAction.Add, item));
+#endif
+
 				return item;
 			}
 
@@ -5390,6 +5469,11 @@
 				list.RemoveAt (index);
 
 #if NET_2_0
+				//UIA Framework event: Item Removed 
+				OnUIACollectionChangedEvent (new CollectionChangeEventArgs (CollectionChangeAction.Remove, item));
+#endif
+
+#if NET_2_0
 				if (is_main_collection) {
 					item.Owner = null;
 					if (item.Group != null)
@@ -5455,6 +5539,11 @@
 #endif
 
 				list.Add (value);
+
+#if NET_2_0
+				//UIA Framework event: Item Added
+				OnUIACollectionChangedEvent (new CollectionChangeEventArgs (CollectionChangeAction.Add, value));
+#endif
 			}
 
 			void CollectionChanged (bool sort)
@@ -6039,5 +6128,62 @@
 		{
 		}
 #endif
+
+#if NET_2_0
+		
+		#region UIA Framework: Methods, Properties and Events
+		
+		static object UIAMultiSelectChangedEvent = new object ();
+		static object UIAViewChangedEvent = new object ();
+
+		internal int UIAColumns {
+			get { return cols; }
+		}
+
+		internal ListViewGroup UIADefaultListViewGroup 
+		{
+			get { return groups.DefaultGroup; }
+		}
+
+		internal ScrollBar UIAHScrollBar {
+			get { return h_scroll; }
+		}
+
+		internal ScrollBar UIAVScrollBar {
+			get { return v_scroll; }
+		}
+
+		internal event EventHandler UIAMultiSelectChanged {
+			add { Events.AddHandler (UIAMultiSelectChangedEvent, value); }
+			remove { Events.RemoveHandler (UIAMultiSelectChangedEvent, value); }
+		}
+
+		internal event EventHandler UIAViewChanged {
+			add { Events.AddHandler (UIAViewChangedEvent, value); }
+			remove { Events.RemoveHandler (UIAViewChangedEvent, value); }
+		}
+
+		internal Rectangle UIAGetHeaderBounds (ListViewGroup group)
+		{
+			return group.HeaderBounds;
+		}
+
+		private void OnUIAMultiSelectChanged ()
+		{
+			EventHandler eh = (EventHandler) Events [UIAMultiSelectChangedEvent];
+			if (eh != null)
+				eh (this, EventArgs.Empty);
+		}
+
+		private void OnUIAViewChanged ()
+		{
+			EventHandler eh = (EventHandler) Events [UIAViewChangedEvent];
+			if (eh != null)
+				eh (this, EventArgs.Empty);
+		}
+
+		#endregion // UIA Framework: Methods, Properties and Events
+
+#endif
 	}
 }
