Index: class/Managed.Windows.Forms/System.Windows.Forms/ListBox.cs
===================================================================
--- class/Managed.Windows.Forms/System.Windows.Forms/ListBox.cs	(revision 111121)
+++ class/Managed.Windows.Forms/System.Windows.Forms/ListBox.cs	(working copy)
@@ -2383,7 +2383,23 @@
 
 			private ListBox owner;
 			internal ArrayList object_items = new ArrayList ();
+			
+			#region UIA Framework Events 
+#if NET_2_0
+			//NOTE:
+			//	We are using Reflection to add/remove internal events.
+			//	Class ListProvider uses the events.
+			//
+			//Event used to generate UIA StructureChangedEvent
+			static object CollectionChangedEvent = new object ();
 
+			internal event CollectionChangeEventHandler CollectionChanged {
+				add { owner.Events.AddHandler (CollectionChangedEvent, value); }
+				remove { owner.Events.RemoveHandler (CollectionChangedEvent, value); }
+			}
+#endif
+			#endregion
+
 			public ObjectCollection (ListBox owner)
 			{
 				this.owner = owner;
@@ -2424,8 +2440,27 @@
 						throw new ArgumentOutOfRangeException ("Index of out range");
 					if (value == null)
 						throw new ArgumentNullException ("value");
+						
+#if NET_2_0
+					//UIA Framework event: Item Removed
+					CollectionChangeEventHandler eh
+						= (CollectionChangeEventHandler) owner.Events [CollectionChangedEvent];
+					if (eh != null)
+						eh (owner,
+						    new CollectionChangeEventArgs (CollectionChangeAction.Remove,
+						                                   index));
+#endif
 
 					object_items[index] = value;
+					
+#if NET_2_0
+					//UIA Framework event: Item Added
+					if (eh != null)
+						eh (owner, 
+						    new CollectionChangeEventArgs (CollectionChangeAction.Add,
+						                                   index));
+#endif					
+
 					owner.CollectionChanged ();
 				}
 			}
@@ -2489,6 +2524,16 @@
 				owner.selected_indices.Clear ();
 				object_items.Clear ();
 				owner.CollectionChanged ();
+
+#if NET_2_0
+				//UIA Framework event: Items list cleared
+				CollectionChangeEventHandler eh 
+					= (CollectionChangeEventHandler) owner.Events [CollectionChangedEvent];
+				if (eh != null)
+					eh (owner,
+					    new CollectionChangeEventArgs (CollectionChangeAction.Refresh,
+					                                   -1));
+#endif
 			}
 			public bool Contains (object value)
 			{
@@ -2549,6 +2594,15 @@
 				object_items.Insert (index, item);
 				owner.CollectionChanged ();
 				owner.EndUpdate ();
+				
+#if NET_2_0
+				//UIA Framework event: Item Added
+				CollectionChangeEventHandler eh 
+					= (CollectionChangeEventHandler) owner.Events [CollectionChangedEvent];
+				if (eh != null)
+					eh (owner,
+					    new CollectionChangeEventArgs (CollectionChangeAction.Add, index));
+#endif
 			}
 
 			public void Remove (object value)
@@ -2567,6 +2621,16 @@
 				owner.selected_indices.Remove (index);
 				object_items.RemoveAt (index);
 				owner.CollectionChanged ();
+				
+#if NET_2_0
+				//UIA Framework event: Item Removed
+				CollectionChangeEventHandler eh
+					= (CollectionChangeEventHandler) owner.Events [CollectionChangedEvent];
+				if (eh != null)
+					eh (owner,
+					    new CollectionChangeEventArgs (CollectionChangeAction.Remove,
+					                                   index));
+#endif
 			}
 			#endregion Public Methods
 
@@ -2578,6 +2642,17 @@
 
 				int cnt = object_items.Count;
 				object_items.Add (item);
+
+#if NET_2_0
+				//UIA Framework event: Item Added
+				CollectionChangeEventHandler eh
+					= (CollectionChangeEventHandler) owner.Events [CollectionChangedEvent];
+				if (eh != null)
+					eh (owner,
+					    new CollectionChangeEventArgs (CollectionChangeAction.Add,
+					                                   cnt));
+#endif
+
 				return cnt;
 			}
 
