Index: atk/Makefile.am
===================================================================
--- atk/Makefile.am	(revision 99178)
+++ atk/Makefile.am	(working copy)
@@ -8,6 +8,7 @@
 sources = 
 
 customs =			\
+	Global.custom		\
 	Object.custom		\
 	Util.custom
 
Index: atk/Global.custom
===================================================================
--- atk/Global.custom	(revision 0)
+++ atk/Global.custom	(revision 0)
@@ -0,0 +1,29 @@
+// Global.custom - Atk Global class customizations
+//
+// Author: Andres G. Aragoneses <aaragoneses@novell.com>
+//
+// Copyright (c) 2008 Novell, Inc.
+//
+// This code is inserted after the automatically generated code.
+//
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of version 2 of the Lesser GNU General 
+// Public License as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this program; if not, write to the
+// Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+// Boston, MA 02111-1307, USA.
+		
+		[DllImport("atksharpglue-2")]
+		static extern void atksharp_global_launch_atkbridge ();
+		
+		public static void LaunchAtkBridge() {
+			atksharp_global_launch_atkbridge ();
+		}
Index: atk/glue/global.c
===================================================================
--- atk/glue/global.c	(revision 0)
+++ atk/glue/global.c	(revision 0)
@@ -0,0 +1,53 @@
+/* global.c : Glue for custom methods of AtkGlobal
+ *
+ * Author: Andres G. Aragoneses  <aaragoneses@novell.com>
+ * 
+ * Copyright (c) 2008 Novell, Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of version 2 of the Lesser GNU General 
+ * Public License as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+#include <stdio.h>
+#include "gmodule.h"
+#include <atk/atk.h>
+
+/* The below line isn't really right -- normally gtk will build the path */
+#define ATK_BRIDGE_PATH "/usr/lib/gtk-2.0/modules/libatk-bridge.so"
+
+void atksharp_global_launch_atkbridge ();
+
+void
+atksharp_global_launch_atkbridge ()
+{
+  GModule *bridge;
+  void (*gnome_accessibility_module_init)();
+  
+  bridge = g_module_open(ATK_BRIDGE_PATH, G_MODULE_BIND_LOCAL|G_MODULE_BIND_LAZY);
+
+  if (!bridge)
+  {
+    fprintf(stderr, "Couldn't load atk-bridge.so\n");
+    exit(1);
+  }
+
+  if (!g_module_symbol(bridge, "gnome_accessibility_module_init", (gpointer *)&gnome_accessibility_module_init))
+  {
+    fprintf(stderr, "Couldn't find gnome_accessibility_module_init\n");    
+    exit(1);
+  }
+
+  (*gnome_accessibility_module_init)();
+}
+
Index: atk/glue/object.c
===================================================================
--- atk/glue/object.c	(revision 99178)
+++ atk/glue/object.c	(working copy)
@@ -1,4 +1,4 @@
-/* util.c : Glue for overriding vms of AtkObject
+/* object.c : Glue for overriding vms of AtkObject
  *
  * Author: Andres G. Aragoneses  <aaragoneses@novell.com>
  * 
Index: atk/glue/Makefile.am
===================================================================
--- atk/glue/Makefile.am	(revision 99178)
+++ atk/glue/Makefile.am	(working copy)
@@ -3,6 +3,7 @@
 libatksharpglue_2_la_LDFLAGS = -module -avoid-version -no-undefined
 
 libatksharpglue_2_la_SOURCES = 	\
+	global.c		\
 	object.c		\
 	util.c			\
 	vmglueheaders.h
