Index: class/Managed.Windows.Forms/System.Windows.Forms/Control.cs
===================================================================
--- class/Managed.Windows.Forms/System.Windows.Forms/Control.cs	(revision 112704)
+++ class/Managed.Windows.Forms/System.Windows.Forms/Control.cs	(working copy)
@@ -5933,9 +5933,9 @@
 				Invalidate ();
 			}
 		}
-
 		#endregion
 
+
 		#region OnXXX methods
 #if NET_2_0
 		protected virtual void OnAutoSizeChanged (EventArgs e)
@@ -6697,6 +6697,7 @@
 		static object ValidatingEvent = new object ();
 		static object VisibleChangedEvent = new object ();
 
+
 #if NET_2_0
 		[Browsable (false)]
 		[EditorBrowsable (EditorBrowsableState.Never)]
@@ -7090,5 +7091,160 @@
 		}
 
 		#endregion	// Events
+
+		#region UIA Framework: Events, Fields and Methods.
+#if NET_2_0
+		//NOTE:
+		//	We are using Reflection to add/remove internal events and 
+		//	get the internal fields.
+		//
+		//ToolTipHookup/ToolTipUnhookup:
+		//	Indicates a ToolTip has been unassociated/associated with the 
+		//	control, used to generate the following Automation Events:
+		//	ChildAdded, ToolTipClosedEvent and ToolTipOpenedEvent
+		//	and to get Automation HelpTextProperty property.
+		//ErrorProviderHookup/ErrorProviderUnhookup:
+		//	Indicates a ToolTip has been unassociated/associated with the 
+		//	control, used to generate the following Automation Events:
+		//	ChildAdded, ToolTipClosedEvent and ToolTipOpenedEvent
+		//	and to get Automation HelpTextProperty property.
+		internal ToolTip uia_tooltip;
+		internal ErrorProvider uia_errorprovider;
+		internal HelpProvider uia_helpprovider;
+
+		static object UIAToolTipHookupEvent = new object ();
+		static object UIAToolTipUnhookupEvent = new object ();
+		static object UIAErrorProviderHookupEvent = new object ();
+		static object UIAErrorProviderUnhookupEvent = new object ();
+		static object UIAHelpProviderHookupEvent = new object ();
+		static object UIAHelpProviderUnhookupEvent = new object ();
+
+		internal event EventHandler UIAToolTipHookedUp {
+			add { Events.AddHandler (UIAToolTipHookupEvent, value); }
+			remove { Events.RemoveHandler (UIAToolTipHookupEvent, value); }
+		}
+
+		internal event EventHandler UIAToolTipUnhookedUp {
+			add { Events.AddHandler (UIAToolTipUnhookupEvent, value); }
+			remove { Events.RemoveHandler (UIAToolTipUnhookupEvent, value); }
+		}
+
+		internal event ControlEventHandler UIAErrorProviderHookedUp {
+			add { Events.AddHandler (UIAErrorProviderHookupEvent, value); }
+			remove { Events.RemoveHandler (UIAErrorProviderHookupEvent, value); }
+		}
+
+		internal event ControlEventHandler UIAErrorProviderUnhookedUp {
+			add { Events.AddHandler (UIAErrorProviderUnhookupEvent, value); }
+			remove { Events.RemoveHandler (UIAErrorProviderUnhookupEvent, value); }
+		}
+
+		internal event EventHandler UIAHelpProviderHookedUp {
+			add { Events.AddHandler (UIAHelpProviderHookupEvent, value); }
+			remove { Events.RemoveHandler (UIAHelpProviderHookupEvent, value); }
+		}
+
+		internal event EventHandler UIAHelpProviderUnhookedUp {
+			add { Events.AddHandler (UIAHelpProviderUnhookupEvent, value); }
+			remove { Events.RemoveHandler (UIAHelpProviderUnhookupEvent, value); }
+		}
+
+		internal ToolTip UIAToolTip {
+			get { return uia_tooltip; }
+			set { 
+				if (uia_tooltip == value)
+					return;
+
+				uia_tooltip = value;
+				if (uia_tooltip != null)
+					OnUIAToolTipHookUp ();
+				else if (uia_tooltip == null)
+					OnUIAToolTipUnhookUp ();
+			}
+		}
+
+		internal ErrorProvider UIAErrorProvider {
+			get { return uia_errorprovider; }
+		}
+
+		internal HelpProvider UIAHelpProvider {
+			get { return uia_helpprovider; } 
+			set {
+				if (uia_helpprovider == value)
+					return;
+
+				uia_helpprovider = value;
+				if (uia_helpprovider != null)
+					OnUIAHelpProviderHookUp ();
+				else if (uia_helpprovider == null)
+					OnUIAHelpProviderUnhookUp ();
+			}
+		}
+
+		internal void OnUIAToolTipHookUp ()
+		{
+			EventHandler eh = (EventHandler) Events [UIAToolTipHookupEvent];
+			if (eh != null)
+				eh (this, EventArgs.Empty);
+		}
+
+		internal void OnUIAToolTipUnhookUp ()
+		{
+			EventHandler eh = (EventHandler) Events [UIAToolTipUnhookupEvent];
+			if (eh != null)
+				eh (this, EventArgs.Empty);
+		}
+
+		internal void OnUIAErrorProviderHookUp (ControlEventArgs args)
+		{
+			ControlEventHandler eh
+				= (ControlEventHandler) Events [UIAErrorProviderHookupEvent];
+			if (eh != null)
+				eh (this, args);
+		}
+
+		internal void OnUIAErrorProviderUnhookedUp (ControlEventArgs args)
+		{
+			ControlEventHandler eh
+				= (ControlEventHandler) Events [UIAErrorProviderUnhookupEvent];
+			if (eh != null)
+				eh (this, args);
+		}
+
+		internal void OnUIAHelpProviderHookUp ()
+		{
+			EventHandler eh = (EventHandler) Events [UIAHelpProviderHookupEvent];
+			if (eh != null)
+				eh (this, EventArgs.Empty);
+		}
+
+		internal void OnUIAHelpProviderUnhookUp ()
+		{
+			EventHandler eh = (EventHandler) Events [UIAHelpProviderUnhookupEvent];
+			if (eh != null)
+				eh (this, EventArgs.Empty);
+		}
+
+		internal void UIAErrorProviderHookUp (ErrorProvider errorProvider, UserControl control)
+		{
+			if (uia_errorprovider == errorProvider)
+				return;
+
+			uia_errorprovider = errorProvider;
+			OnUIAErrorProviderHookUp (new ControlEventArgs (control));
+		}
+
+		internal void UIAErrorProviderUnhookUp (ErrorProvider errorProvider, UserControl control)
+		{
+			if (uia_errorprovider != errorProvider)
+				return;
+
+			uia_errorprovider = null;
+			OnUIAErrorProviderUnhookedUp (new ControlEventArgs (control));
+		}
+#endif
+		#endregion
+
+
 	}
 }
Index: class/Managed.Windows.Forms/System.Windows.Forms/ToolTip.cs
===================================================================
--- class/Managed.Windows.Forms/System.Windows.Forms/ToolTip.cs	(revision 112704)
+++ class/Managed.Windows.Forms/System.Windows.Forms/ToolTip.cs	(working copy)
@@ -77,6 +77,15 @@
 			#region ToolTipWindow Class Local Variables
 			private Control associated_control;
 			#endregion	// ToolTipWindow Class Local Variables
+			
+			#region UIA Framework: Property
+#if NET_2_0
+			//Property used to set ToolTip's visibility in Control
+			internal Control UIAControl {
+				get { return associated_control; }
+			}
+#endif
+			#endregion UIA Framework
 
 			#region ToolTipWindow Class Constructor
 			internal ToolTipWindow() {
@@ -293,6 +302,17 @@
 			timer = new Timer();
 			timer.Enabled = false;
 			timer.Tick +=new EventHandler(timer_Tick);
+			
+
+#if NET_2_0
+			//UIA Framework: Event used to indicate that ToolTip is shown
+			tooltip_window.VisibleChanged += delegate (object sender, EventArgs args) {
+				if (tooltip_window.Visible == true)
+					OnUIAToolTipShown (new ControlEventArgs (tooltip_window.UIAControl));
+				else 
+					OnUIAToolTipHidden (new ControlEventArgs (tooltip_window.UIAControl));
+			};
+#endif
 		}
 
 		public ToolTip(System.ComponentModel.IContainer cont) : this() {
@@ -504,12 +524,16 @@
 
 		public void RemoveAll() {
 			tooltip_strings.Clear();
+#if NET_2_0
+			//UIA Framework: ToolTip isn't associated anymore
+			foreach (Control control in controls)
+				control.UIAToolTip = null;
+#endif
 			controls.Clear();
 		}
 
 		public void SetToolTip(Control control, string caption) {
 			tooltip_strings[control] = caption;
-
 			// no need for duplicates
 			if (!controls.Contains(control)) {
 				control.MouseEnter += new EventHandler(control_MouseEnter);
@@ -517,6 +541,10 @@
 				control.MouseLeave += new EventHandler(control_MouseLeave);
 				control.MouseDown += new MouseEventHandler (control_MouseDown);
 				controls.Add(control);
+#if NET_2_0
+				//UIA Framework: Associate ToolTip with Control
+				control.UIAToolTip = this;
+#endif
 			}
 			
 			// if SetToolTip is called from a control and the mouse is currently over that control,
@@ -657,6 +685,11 @@
 
 				tooltip_strings.Clear();
 				
+#if NET_2_0
+				//UIA Framework: ToolTip isn't associated anymore
+				foreach (Control control in controls)
+					control.UIAToolTip = null;
+#endif
 				controls.Clear();
 			}
 		}
@@ -697,6 +730,8 @@
 				control.MouseLeave += new EventHandler (control_MouseLeave);
 				control.MouseDown += new MouseEventHandler (control_MouseDown);
 				controls.Add (control);
+				//UIA Framework: Associate ToolTip with Control
+				control.UIAToolTip = this;
 			}
 		}
 
@@ -819,7 +854,7 @@
 #endif
 				ThemeEngine.Current.DrawToolTip (e.Graphics, e.Bounds, tooltip_window);
 		}
-
+		
 		private bool MouseInControl (Control control, bool fuzzy) {
 			Point	m;
 			Point	c;
@@ -934,5 +969,40 @@
 			remove { Events.RemoveHandler (DrawEvent, value); }
 		}
 		#endregion
+
+		#region UIA Framework: Events, Delegates and Methods
+#if NET_2_0
+
+		static object UIAToolTipShownEvent = new object ();
+		static object UIAToolTipHiddenEvent = new object ();
+
+		internal event ControlEventHandler UIAToolTipShown {
+			add { Events.AddHandler (UIAToolTipShownEvent, value); }
+			remove { Events.RemoveHandler (UIAToolTipShownEvent, value); }
+		}
+
+		internal event ControlEventHandler UIAToolTipHidden {
+			add { Events.AddHandler (UIAToolTipHiddenEvent, value); }
+			remove { Events.RemoveHandler (UIAToolTipHiddenEvent, value); }
+		}
+
+		internal void OnUIAToolTipHidden (ControlEventArgs args)
+		{
+			ControlEventHandler eh 
+				= (ControlEventHandler) Events [UIAToolTipHiddenEvent];
+			if (eh != null)
+				eh (this, args);
+		}
+
+		internal void OnUIAToolTipShown (ControlEventArgs args)
+		{
+			ControlEventHandler eh 
+				= (ControlEventHandler) Events [UIAToolTipShownEvent];
+			if (eh != null)
+				eh (this, args);
+		}
+
+#endif
+		#endregion
 	}
 }
Index: class/Managed.Windows.Forms/System.Windows.Forms/HelpProvider.cs
===================================================================
--- class/Managed.Windows.Forms/System.Windows.Forms/HelpProvider.cs	(revision 112704)
+++ class/Managed.Windows.Forms/System.Windows.Forms/HelpProvider.cs	(working copy)
@@ -100,6 +100,16 @@
 			controls = new Hashtable();
 			tooltip = new ToolTip.ToolTipWindow();
 
+#if NET_2_0
+			//UIA Framework: Event used to indicate that ToolTip is shown
+			tooltip.VisibleChanged += delegate (object sender, EventArgs args) {
+				if (tooltip.Visible == true)
+					OnUIAToolTipShown (new ControlEventArgs (UIAControl));
+				else 
+					OnUIAToolTipHidden (new ControlEventArgs (UIAControl));
+			};
+#endif
+
 			HideToolTipHandler = new EventHandler(HideToolTip);
 			HideToolTipKeyHandler = new KeyPressEventHandler(HideToolTipKey);
 			HideToolTipMouseHandler = new MouseEventHandler(HideToolTipMouse);
@@ -181,6 +191,11 @@
 			} else {
 				hp.Show = false;
 			}
+
+#if NET_2_0
+			//UIA Framework: Associate/disassociate HelpProvider with control
+			ctl.UIAHelpProvider = hp.Show == true ? this : null;
+#endif
 		}
 
 		public virtual void SetHelpKeyword(Control ctl, string keyword) {
@@ -193,10 +208,20 @@
 
 		public virtual void SetHelpString(Control ctl, string helpString) {
 			GetHelpProperty(ctl).Text = helpString;
+#if NET_2_0
+			//UIA Framework: Disassociate HelpProvider with control
+			if (helpString == string.Empty || helpString == null)
+				ctl.UIAHelpProvider = null;
+#endif
 		}
 
 		public virtual void SetShowHelp(Control ctl, bool value) {
 			GetHelpProperty(ctl).Show = value;
+
+#if NET_2_0
+			//UIA Framework: Associate/disassociate HelpProvider with control
+			ctl.UIAHelpProvider = GetHelpProperty(ctl).Show == true ? this : null;
+#endif
 		}
 
 		public override string ToString() {
@@ -213,6 +238,11 @@
 			if (hp == null) {
 				hp = new HelpProperty(this, control);
 				controls[control] = hp;
+
+#if NET_2_0
+				//UIA Framework: Associate HelpProvider with control
+				control.UIAHelpProvider = this;
+#endif
 			}
 
 			return hp;
@@ -254,6 +284,11 @@
 
 			control = (Control)sender;
 
+#if NET_2_0
+			//UIA Framework: Associates requested control with internal variable to generate event
+			UIAControl = control;
+#endif
+
 			if (GetHelpProperty(control).Text == null) {
 				return;
 			}
@@ -291,5 +326,46 @@
 			e.Handled = true;
 		}
 		#endregion	// Private Methods
+
+		#region UIA Framework: Events, Delegates and Methods
+#if NET_2_0
+
+		static object UIAToolTipShownEvent = new object ();
+		static object UIAToolTipHiddenEvent = new object ();
+		private Control uia_control;
+		
+		private Control UIAControl {
+			get { return uia_control; }
+			set { uia_control = value; }
+		}
+
+		internal event ControlEventHandler UIAToolTipShown {
+			add { Events.AddHandler (UIAToolTipShownEvent, value); }
+			remove { Events.RemoveHandler (UIAToolTipShownEvent, value); }
+		}
+
+		internal event ControlEventHandler UIAToolTipHidden {
+			add { Events.AddHandler (UIAToolTipHiddenEvent, value); }
+			remove { Events.RemoveHandler (UIAToolTipHiddenEvent, value); }
+		}
+
+		internal void OnUIAToolTipHidden (ControlEventArgs args)
+		{
+			ControlEventHandler eh 
+				= (ControlEventHandler) Events [UIAToolTipHiddenEvent];
+			if (eh != null)
+				eh (this, args);
+		}
+
+		internal void OnUIAToolTipShown (ControlEventArgs args)
+		{
+			ControlEventHandler eh 
+				= (ControlEventHandler) Events [UIAToolTipShownEvent];
+			if (eh != null)
+				eh (this, args);
+		}
+
+#endif
+		#endregion
 	}
 }
Index: class/Managed.Windows.Forms/System.Windows.Forms/ErrorProvider.cs
===================================================================
--- class/Managed.Windows.Forms/System.Windows.Forms/ErrorProvider.cs	(revision 112704)
+++ class/Managed.Windows.Forms/System.Windows.Forms/ErrorProvider.cs	(working copy)
@@ -79,10 +79,30 @@
 				window.Width = ep.icon.Width;
 				window.Height = ep.icon.Height;
 
+#if NET_2_0
+				//UIA Framework: Associates ErrorProvider with Control
+				window.VisibleChanged += delegate (object sender, EventArgs args) {
+					if (((UserControl) sender).Visible == false)
+						control.UIAErrorProviderUnhookUp (ep, window);
+					else
+						control.UIAErrorProviderHookUp (ep, window);
+				};
+#endif
+
 				if (ep.container != null && control.Parent != null) {
+#if NET_2_0
+					//UIA Framework: Associates ErrorProvider with Control
+					control.UIAErrorProviderHookUp (ep, window);
+#endif
+					
 					ep.container.Controls.Add(window);
 					ep.container.Controls.SetChildIndex (window, control.Parent.Controls.IndexOf (control) + 1);
 				} else if (control.Parent != null) {
+#if NET_2_0
+					//UIA Framework: Associates ErrorProvider with Control
+					control.UIAErrorProviderHookUp (ep, window);
+#endif
+
 					control.Parent.Controls.Add(window);
 					control.Parent.Controls.SetChildIndex(window, control.Parent.Controls.IndexOf (control) + 1);
 				}
@@ -223,6 +243,10 @@
 					ep.tooltip.Height = size.Height;
 					ep.tooltip.Text = text;
 
+#if NET_2_0
+					ep.UIAControl = control;
+#endif
+
 					if ((pt.X + size.Width) < SystemInformation.WorkingArea.Width) {
 						ep.tooltip.Left = pt.X;
 					} else {
@@ -312,7 +336,49 @@
 		private object tag;
 #endif
 		#endregion	// Local Variables
+	
+		#region UIA Framework: Events, Delegates and Methods
+#if NET_2_0
 
+		static object UIAToolTipShownEvent = new object ();
+		static object UIAToolTipHiddenEvent = new object ();
+
+		internal event ControlEventHandler UIAToolTipShown {
+			add { Events.AddHandler (UIAToolTipShownEvent, value); }
+			remove { Events.RemoveHandler (UIAToolTipShownEvent, value); }
+		}
+
+		internal event ControlEventHandler UIAToolTipHidden {
+			add { Events.AddHandler (UIAToolTipHiddenEvent, value); }
+			remove { Events.RemoveHandler (UIAToolTipHiddenEvent, value); }
+		}
+
+		internal void OnUIAToolTipHidden (ControlEventArgs args) 
+		{
+			ControlEventHandler eh 
+				= (ControlEventHandler) Events [UIAToolTipHiddenEvent];
+			if (eh != null)
+				eh (this, args);
+		}
+
+		internal void OnUIAToolTipShown (ControlEventArgs args) 
+		{
+			ControlEventHandler eh 
+				= (ControlEventHandler) Events [UIAToolTipShownEvent];
+			if (eh != null)
+				eh (this, args);
+		}
+
+		private Control uia_control;
+		
+		internal Control UIAControl {
+			set { uia_control = value; }
+			get { return uia_control; }
+		}
+
+#endif
+		#endregion
+
 		#region Public Constructors
 		public ErrorProvider()
 		{
@@ -323,6 +389,18 @@
 
 			icon = ResourceImageLoader.GetIcon ("errorProvider.ico");
 			tooltip = new ToolTip.ToolTipWindow();
+
+#if NET_2_0
+			//UIA Framework: Event used to indicate the ToolTip is shown/hidden.
+			tooltip.VisibleChanged += delegate (object sender, EventArgs args) {
+				Console.WriteLine ("Visible changed: {0}", tooltip.Visible);
+				if (tooltip.Visible == true)
+					OnUIAToolTipShown (new ControlEventArgs (UIAControl));
+				else if (tooltip.Visible == false)
+					OnUIAToolTipHidden (new ControlEventArgs (UIAControl));
+			};
+#endif
+
 		}
 
 		public ErrorProvider(ContainerControl parentControl) : this ()
