Index: class/Managed.Windows.Forms/System.Windows.Forms/X11DesktopColors.cs
===================================================================
--- class/Managed.Windows.Forms/System.Windows.Forms/X11DesktopColors.cs	(revision 105916)
+++ class/Managed.Windows.Forms/System.Windows.Forms/X11DesktopColors.cs	(working copy)
@@ -28,6 +28,7 @@
 using System.Drawing;
 using System.Runtime.InteropServices;
 using System.IO;
+using System;
 
 namespace System.Windows.Forms {
 	internal class X11DesktopColors {
@@ -97,7 +98,7 @@
 					GtkStyleStruct	style;
 
 					try {
-						gtk_init_check (IntPtr.Zero, IntPtr.Zero);
+						GtkInit();
 						//dispmgr =  gdk_display_manager_get ();
 						//gdkdisplay =  gdk_display_manager_get_default_display (dispmgr);
 
@@ -145,6 +146,27 @@
 				}
 			}
 		}
+
+		static void GtkInit ()
+		{
+//no SetEnv in 1.x
+#if NET_2_0
+			// Prevent ATK bridge initalization, as it causes
+			// conflicts with the UIA<->ATK bridge.
+			// A better solution would probably be to get these
+			// values out-of-process.  See details/discussion here:
+			// https://bugzilla.novell.com/show_bug.cgi?id=375987
+			string gtk_modules_env_var = "GTK_MODULES";
+			string gtk_modules_env_var_content = Environment.GetEnvironmentVariable (gtk_modules_env_var);
+			Environment.SetEnvironmentVariable (gtk_modules_env_var, string.Empty);
+#endif
+
+			gtk_init_check (IntPtr.Zero, IntPtr.Zero);
+
+#if NET_2_0
+			Environment.SetEnvironmentVariable (gtk_modules_env_var, gtk_modules_env_var_content);
+#endif
+		}
 		#endregion	// Constructors
 
 		#region Properties
Index: class/Managed.Windows.Forms/System.Windows.Forms/Application.cs
===================================================================
--- class/Managed.Windows.Forms/System.Windows.Forms/Application.cs	(revision 105916)
+++ class/Managed.Windows.Forms/System.Windows.Forms/Application.cs	(working copy)
@@ -158,8 +158,55 @@
 		{
 		}
 
+#if NET_2_0
+		static Application ()
+		{
+			// Attempt to load UIA support for winforms
+			// UIA support requires .NET 2.0
+			InitializeUIAutomation ();
+		}
+#endif
+
 		#region Private Methods
 
+#if NET_2_0
+		private static void InitializeUIAutomation ()
+		{
+			// Initialize the UIAutomationWinforms FormListener
+			// class, which signs up for the PreRun and FormAdded
+			// events so that it can provide a11y support for MWF.
+			string uia_winforms_assembly =
+				"UIAutomationWinforms, Version=1.0.0.0, Culture=neutral, PublicKeyToken=f4ceacb585d99812";
+			string listener_type_name =
+				"Mono.UIAutomation.Winforms.FormListener";
+			string init_method_name = "Initialize";
+			
+			Assembly mwf_providers = null;
+			try {
+				mwf_providers =
+					Assembly.Load (uia_winforms_assembly);
+			} catch { }
+			
+			if (mwf_providers == null)
+				return;
+			
+			try {
+				Type listener_type =
+					mwf_providers.GetType (listener_type_name,
+						false);
+				if (listener_type == null)
+					return;
+				MethodInfo init_method =
+					listener_type.GetMethod (init_method_name,
+						BindingFlags.Static | BindingFlags.Public);
+				if (init_method == null)
+					return;
+				init_method.Invoke (null,
+					new object [] {});
+			} catch { }
+		}
+#endif
+		
 		internal static void CloseForms (Thread thread)
 		{
 			#if DebugRunLoop
@@ -685,6 +732,11 @@
 		public static void Run (ApplicationContext context)
 		{
 #if NET_2_0
+			// Signal that the Application loop is about to run.
+			// This allows UIA to initialize a11y support for MWF
+			// before the main loop begins.
+			if (PreRun != null)
+				PreRun (null, EventArgs.Empty);
 			// If a sync context hasn't been created by now, create
 			// a default one
 			if (SynchronizationContext.Current == null)
@@ -974,8 +1026,12 @@
 
 		public static event EventHandler ThreadExit;
 		public static event ThreadExceptionEventHandler ThreadException;
+		
+#if NET_2_0
+		// These are used externally by the UIA framework
+		internal static event EventHandler FormAdded;
+		internal static event EventHandler PreRun;
 
-#if NET_2_0
 		[EditorBrowsable (EditorBrowsableState.Advanced)]
 		public static event EventHandler EnterThreadModal;
 
@@ -1009,6 +1065,15 @@
 		{
 			lock (forms)
 				forms.Add (f);
+#if NET_2_0
+			// Signal that a Form has been added to this
+			// Application. Used by UIA to detect new Forms that
+			// need a11y support. This event may be fired even if
+			// the form has already been added, so clients should
+			// account for that when handling this signal.
+			if (FormAdded != null)
+				FormAdded (f, null);
+#endif
 		}
 		
 		internal static void RemoveForm (Form f)
