Index: class/Managed.Windows.Forms/System.Windows.Forms/X11DesktopColors.cs
===================================================================
--- class/Managed.Windows.Forms/System.Windows.Forms/X11DesktopColors.cs	(revision 96906)
+++ class/Managed.Windows.Forms/System.Windows.Forms/X11DesktopColors.cs	(working copy)
@@ -86,6 +86,13 @@
 
 		#region Constructors
 		static X11DesktopColors() {
+			// Prevent ATK bridge initalization, as it causes
+			// conflicts with the UIA<->ATK bridge.
+			// A better solution would probably be to get these
+			// values out-of-process.  See details/discussion here:
+			// https://bugzilla.novell.com/show_bug.cgi?id=375987
+			Environment.SetEnvironmentVariable ("GTK_MODULES", string.Empty);
+			
 			FindDesktopEnvironment();
 			
 			switch(desktop) {
Index: class/Managed.Windows.Forms/System.Windows.Forms/Application.cs
===================================================================
--- class/Managed.Windows.Forms/System.Windows.Forms/Application.cs	(revision 96906)
+++ class/Managed.Windows.Forms/System.Windows.Forms/Application.cs	(working copy)
@@ -153,10 +153,42 @@
 		private static ToolStrip keyboard_capture;
 		private static VisualStyleState visual_style_state = VisualStyleState.ClientAndNonClientAreasEnabled;
 #endif
+		
+		private const string UIAutomationWinformsAssembly =
+			"UIAutomationWinforms, Version=1.0.0.0, Culture=neutral, PublicKeyToken=f4ceacb585d99812";
+		private const string UIAFormListenerType =
+			"Mono.UIAutomation.Winforms.FormListener";
+		private const string UIAFormListnerInitializeMethod =
+			"Initialize";
 
 		private Application ()
 		{
 		}
+		
+		// TODO: Any #if checks required here?
+		static Application ()
+		{
+			// TODO: Check if accessibility is turned on first
+			Assembly mwfProviders = null;
+			try {
+				mwfProviders =
+					Assembly.Load (UIAutomationWinformsAssembly);
+			} catch { }
+			
+			if (mwfProviders != null) {
+				// TODO: Need try/catch checks here because of
+				//       uiawinforms loading other assemblies
+				Type listenerType =
+					mwfProviders.GetType (UIAFormListenerType, false);
+				if (listenerType != null) {
+					MethodInfo initMethod =
+						listenerType.GetMethod (UIAFormListnerInitializeMethod, BindingFlags.Static | BindingFlags.Public);
+					if (initMethod != null) {
+						initMethod.Invoke (null, new object [] {});
+					}
+				}
+			}
+		}
 
 		#region Private Methods
 
@@ -671,9 +703,13 @@
 		{
 			Run (new ApplicationContext (mainForm));
 		}
+		
+		internal static event EventHandler PreRun;
 
 		public static void Run (ApplicationContext context)
 		{
+			if (PreRun != null)
+				PreRun (null, new EventArgs ());
 #if NET_2_0
 			// If a sync context hasn't been created by now, create
 			// a default one
@@ -964,6 +1000,8 @@
 
 		public static event EventHandler ThreadExit;
 		public static event ThreadExceptionEventHandler ThreadException;
+		
+		internal static event EventHandler FormAdded;
 
 #if NET_2_0
 		[EditorBrowsable (EditorBrowsableState.Advanced)]
@@ -999,6 +1037,8 @@
 		{
 			lock (forms)
 				forms.Add (f);
+			if (FormAdded != null)
+				FormAdded (f, null);
 		}
 		
 		internal static void RemoveForm (Form f)
 
